store/v2/root/store.go

```diff
@@ -0,0 +1,938 @@
+	return
+}
+
+func (s *Store) Close() error {
+	err := s.stateTxn.Discard()
+	if s.StateCommitmentDB != nil {
+		err = util.CombineErrors(err, s.stateCommitmentTxn.Discard(), "stateCommitmentTxn.Discard also failed")
+	}
+	return err
+}
+
+// Applies store upgrades to the DB contents.
+func (pr *prefixRegistry) migrate(store *Store, upgrades types.StoreUpgrades) error {
+	// Get a view of current state to allow mutation while iterating
+	reader := store.stateDB.Reader()
+	scReader := reader
+	if store.StateCommitmentDB != nil {
+		scReader = store.StateCommitmentDB.Reader()
+	}
+
+	for _, key := range upgrades.Deleted {
+		sst, ix, err := pr.storeInfo(key)
+		if err != nil {
+			return err
+		}
+		if sst != types.StoreTypePersistent {
+			return fmt.Errorf("prefix is for non-persistent substore: %v (%v)", key, sst)
+		}
+		pr.reserved = append(pr.reserved[:ix], pr.reserved[ix+1:]...)
+		delete(pr.StoreSchema, key)
+
+		pfx := substorePrefix(key)
+		subReader := prefixdb.NewPrefixReader(reader, pfx)
+		it, err := subReader.Iterator(nil, nil)
+		if err != nil {
+			return err
+		}
+		for it.Next() {
+			store.stateTxn.Delete(it.Key())
+		}
+		it.Close()
+		if store.StateCommitmentDB != nil {
+			subReader = prefixdb.NewPrefixReader(scReader, pfx)
+			it, err = subReader.Iterator(nil, nil)
+			if err != nil {
+				return err
+			}
+			for it.Next() {
+				store.stateCommitmentTxn.Delete(it.Key())
+			}
+			it.Close()
+		}
+	}
+	for _, rename := range upgrades.Renamed {
+		sst, ix, err := pr.storeInfo(rename.OldKey)
+		if err != nil {
+			return err
+		}
+		if sst != types.StoreTypePersistent {
+			return fmt.Errorf("prefix is for non-persistent substore: %v (%v)", rename.OldKey, sst)
+		}
+		pr.reserved = append(pr.reserved[:ix], pr.reserved[ix+1:]...)
+		delete(pr.StoreSchema, rename.OldKey)
+		err = pr.ReservePrefix(rename.NewKey, types.StoreTypePersistent)
+		if err != nil {
+			return err
+		}
+
+		oldPrefix := substorePrefix(rename.OldKey)
+		newPrefix := substorePrefix(rename.NewKey)
+		subReader := prefixdb.NewPrefixReader(reader, oldPrefix)
+		subWriter := prefixdb.NewPrefixWriter(store.stateTxn, newPrefix)
+		it, err := subReader.Iterator(nil, nil)
+		if err != nil {
+			return err
+		}
+		for it.Next() {
+			subWriter.Set(it.Key(), it.Value())
+		}
+		it.Close()
+		if store.StateCommitmentDB != nil {
+			subReader = prefixdb.NewPrefixReader(scReader, oldPrefix)
+			subWriter = prefixdb.NewPrefixWriter(store.stateCommitmentTxn, newPrefix)
+			it, err = subReader.Iterator(nil, nil)
+			if err != nil {
+				return err
+			}
+			for it.Next() {
+				subWriter.Set(it.Key(), it.Value())
+			}
+			it.Close()
+		}
+	}
+
+	for _, key := range upgrades.Added {
+		err := pr.ReservePrefix(key, types.StoreTypePersistent)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func substorePrefix(key string) []byte {
+	return append(contentPrefix, key...)
+}
+
+// GetKVStore implements BasicRootStore.
+// Returns a substore whose contents
```

Comment:
let's explain that each substore has it's own instance of SMT, but it's using the same physical DB. We also need to update the documentation in `docs/`